plugin/evm/vm.go

```diff
@@ -1023,92 +1019,127 @@ func (vm *VM) initBlockBuilding() error {
 	if err != nil {
 		return err
 	}
 
-	atomicTxGossipHandler = &p2p.ValidatorHandler{
+	atomicTxPullGossipHandler = &p2p.ValidatorHandler{
 		ValidatorSet: vm.validators,
 		Handler: &p2p.ThrottlerHandler{
 			Throttler: p2p.NewSlidingWindowThrottler(throttlingPeriod, throttlingLimit),
-			Handler:   atomicTxGossipHandler,
+			Handler:   atomicTxPullGossipHandler,
 		},
 	}
 
-	atomicTxGossipClient, err := vm.router.RegisterAppProtocol(atomicTxGossipProtocol, atomicTxGossipHandler, vm.validators)
+	atomicTxPullGossipClient, err := vm.router.RegisterAppProtocol(atomicTxPullGossipProtocol, atomicTxPullGossipHandler, vm.validators)
 	if err != nil {
 		return err
 	}
 
 	var (
-		ethTxGossiper    gossip.Gossiper
-		atomicTxGossiper gossip.Gossiper
+		ethTxPullGossiper    gossip.Gossiper
+		atomicTxPullGossiper gossip.Gossiper
 	)
-	ethTxGossiper, err = gossip.NewPullGossiper[GossipEthTx, *GossipEthTx](
+
+	ethTxPullGossiper, err = gossip.NewPullGossiper[GossipEthTx, *GossipEthTx](
 		ethTxGossipConfig,
 		vm.ctx.Log,
 		ethTxPool,
-		ethTxGossipClient,
+		ethTxPullGossipClient,
 		vm.sdkMetrics,
 	)
 	if err != nil {
 		return err
 	}
-	ethTxGossiper = gossip.ValidatorGossiper{
-		Gossiper:   ethTxGossiper,
+
+	ethTxPullGossiper = gossip.ValidatorGossiper{
+		Gossiper:   ethTxPullGossiper,
 		NodeID:     vm.ctx.NodeID,
 		Validators: vm.validators,
 	}
 
 	vm.shutdownWg.Add(1)
 	go func() {
-		gossip.Every(ctx, vm.ctx.Log, ethTxGossiper, gossipFrequency)
+		gossip.Every(ctx, vm.ctx.Log, ethTxPullGossiper, gossipFrequency)
 		vm.shutdownWg.Done()
 	}()
 
-	atomicTxGossiper, err = gossip.NewPullGossiper[GossipAtomicTx, *GossipAtomicTx](
+	atomicTxPullGossiper, err = gossip.NewPullGossiper[GossipAtomicTx, *GossipAtomicTx](
 		atomicTxGossipConfig,
 		vm.ctx.Log,
 		vm.mempool,
-		atomicTxGossipClient,
+		atomicTxPullGossipClient,
 		vm.sdkMetrics,
 	)
 	if err != nil {
 		return err
 	}
-	atomicTxGossiper = gossip.ValidatorGossiper{
-		Gossiper:   atomicTxGossiper,
+
+	atomicTxPullGossiper = gossip.ValidatorGossiper{
+		Gossiper:   atomicTxPullGossiper,
 		NodeID:     vm.ctx.NodeID,
 		Validators: vm.validators,
 	}
 
 	vm.shutdownWg.Add(1)
 	go func() {
-		gossip.Every(ctx, vm.ctx.Log, atomicTxGossiper, gossipFrequency)
+		gossip.Every(ctx, vm.ctx.Log, atomicTxPullGossiper, gossipFrequency)
 		vm.shutdownWg.Done()
 	}()
 
+	gossipStats := NewGossipStats()
+	vm.gossipHandler = NewGossipHandler(vm, gossipStats)
+
+	ethTxPushGossipHandler := ethTxPushGossipHandler{
+		NoOpHandler:   p2p.NoOpHandler{},
+		GossipHandler: vm.gossipHandler,
+		codec:         vm.networkCodec,
+	}
+
+	ethTxPushGossipClient, err := vm.router.RegisterAppProtocol(ethTxPushGossipProtocol, ethTxPushGossipHandler, nil)
+	if err != nil {
+		return err
+	}
+
+	atomicTxPushGossipHandler := atomicTxPushGossipHandler{
+		NoOpHandler:   p2p.NoOpHandler{},
+		GossipHandler: vm.gossipHandler,
+		codec:         vm.networkCodec,
+	}
```

Comment:
Seems there are bunch of duplicated code. I wonder if we can combine them into a single function that works with interfaces and take handlers as parameter to return client. I.e initGossiper(mempool gossip.gossippable, protocol, gossip.config...)

Reply:
I think there should be less duplicated code now that there's only one gossip handler per gossip protocol... previously there was 2x2=4 which was definitely too much...

Reply:
Added function to do this